[Feature]: Manage select options via interface

kjkooistra-youwe opened the pull request:
- Add Select Options management interface via Settings > Data Objects submenu.
- Add selectoptions permission to allow access to management interface.
- Update select fields to easily configure select options via options provider.
- Generate definition file and enum to store select options.
- Introduce OptionsProviderInterface to detect field definitions using provider.
- Update classes rebuild command to also rebuild select options.
- Add PHP and JavaScript reserved words helper classes.
- Add documentation.

Relates to #3409

Future changes
- Add export/import options.
- Restrict editing to admins per select options definition, allowing (regular) user access to specific select options.
- Use reserved words helpers for other components.

kjkooistra-youwe committed:
[Feature]: Manage select options via interface

Add Select Options management interface via Settings > Data Objects submenu.
Add selectoptions permission to allow access to management interface.
Add options source selection to select fields definition to switch between
direct configuration, select options or class / service configuration.
Generate configuration file and enum to store select options.
Introduce OptionsProviderInterface to detect field definitions using provider.
Introduce OptionsProviderTrait to share properties between select types.
Update classes rebuild command to also rebuild select options.
Add PHP and JavaScript reserved words helper classes.
Add and update documentation.
Relates to #3409

jdreesen commented on models/DataObject/SelectOptions/Data/SelectOption.php:
```diff
@@ -0,0 +1,108 @@
+
+    public function hasLabel(): bool
+    {
+        return !empty($this->label) && $this->label !== '0';
```
Since it's a `string`, you could just write (also for the other fields):

```suggestion
        return $this->label !== '';
```

jdreesen commented on bundles/CoreBundle/src/Command/ClassesRebuildCommand.php:
```diff
@@ -145,6 +145,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+            }
+
+            $selectOptionDefinition->save(false);
+        }
```
Do we also need to add this here? https://github.com/pimcore/pimcore/blob/11.x/bundles/CoreBundle/src/Command/ClassesDefinitionsBuildCommand.php

jdreesen commented on lib/Helper/ReservedWordsHelper.php:
```diff
@@ -0,0 +1,103 @@
+     * @return string[]
+     */
+    public function getPhpReservedKeywords(): array {
+        return [
+            'abstract', 'and', 'array', 'as', 'break', 'callable', 'case', 'catch', 'class', 'clone', 'const', 'continue',
+            'declare', 'default', 'die', 'do', 'echo', 'else', 'elseif', 'empty', 'enddeclare', 'endfor', 'endforeach',
+            'endif', 'endswitch', 'endwhile', 'eval', 'exit', 'extends', 'final', 'finally', 'fn', 'for', 'foreach',
+            'function', 'global', 'goto', 'if', 'implements', 'include', 'include_once', 'instanceof', 'insteadof',
+            'interface', 'isset', 'list', 'match', 'namespace', 'new', 'or', 'print', 'private', 'protected', 'public',
+            'readonly', 'require', 'require_once', 'return', 'static', 'switch', 'throw', 'trait', 'try', 'unset', 'use',
+            'var', 'while', 'xor', 'yield', 'yield_from'
+        ];
```
What about defining these as `const`s?

kjkooistra-youwe replied:
Good idea.

jdreesen commented on doc/Development_Documentation/05_Objects/01_Object_Classes/01_Data_Types/77_Select_Options.md:
```diff
@@ -0,0 +1,70 @@
+# Select Options
+
+Select options are predefined sets op options which may be used for (multi)select fields.
```
```suggestion
Select options are predefined sets of options which may be used for (multi)select fields.
```

jdreesen reviewed (commented):
Very nice :) I just did a quick review.

NiklasBr commented on bundles/AdminBundle/src/Controller/Admin/DataObject/ClassController.php:
```diff
@@ -1900,6 +1902,26 @@ public function getBrickUsagesAction(Request $request): Response
         return $this->adminJson($result);
     }
 
+    /**
+     * @Route("/get-select-options-usages", name="getselectoptionsusages", methods={"GET"})
+     */
```
```suggestion
     #[Route("/get-select-options-usages", name: "getselectoptionsusages", methods: ["GET])]
```

NiklasBr replied:
I think as of Pimcore 11 attributes should be used.

kjkooistra-youwe replied:
It made sense to me to keep it consistent with the other routes in this controller. But I guess the difference isn't that big.

kjkooistra-youwe commented:
Resolved merge conflicts.

lukmzig commented:
Hi @kjkooistra-youwe, thank you very much for this improvement, in general we really like it! We have couple of suggestions/questions, here they are:

1. Errors
 - Upon creating/opening existing Select option definition in admin interface an exception is thrown (**Message: trim(): Argument 1 ($string) must be of type string, null given**)
 - **Used by class** seems to be not working currently - probably because of the exception above

2. Questions
- What is the difference between the Display name and Name?

3. Improvements
- If I am correct, the definition for select options is currently stored in the JSON file. This could become performance demanding. We would suggest implementing LocationAwareConfigRepository with the write targets. Then it would be possible to define if these definitions should be stored in the DB or as the yaml file. 

- It would be good to add possibility to localize Display name.  (We should discuss this point in more detail before you would start working on it)

- In the class definition it would be good to add an extra select field (Something like Options provider type). Here we could have 3 options:

  - default option (pre-selected) - Selection Options table would be displayed, so user can add the options manually
  - custom options provider - hide Selection Options table and display **Options Provider Class or Service Name** and **Options Provider Data** fields
  - select options defnition - shows only the Select Options dropdown. Also when you select some option in this drop down it would be probably good that it stays pre-selected on the class definition re-load.
  
Let us know what do you think about these points. Would you have time and motivation to have a look at these? We really appreciate your contribution, thanks agian!

kjkooistra-youwe commented:
@lukmzig Thanks for your findings.

1. Errors
I'll see if I can reproduce and fix it when I have some time.

2. Questions
I copied the (multi)select field options configuration, so the Display Name is the same (the option label).

The enum cases are generated from the option values by default, but that doesn't work for certain special characters (`√`, for example). That's why an additional Name column was added, to be able to configure a valid (or custom) enum case. Also see the `Check` case in the [documentation](https://github.com/pimcore/pimcore/blob/07c97abeda6af172bc980f8d86f0254d162e5cd2/doc/Development_Documentation/05_Objects/01_Object_Classes/01_Data_Types/77_Select_Options.md).

3. Improvements
- Definition: I'll look into it when I have some time. 
- Localization: I saw this in the linked issue, but left it out on purpose, as it would make it even more complex. I would prefer to create the base setup first and then build further on that.
- Option provider type: this does make sense. Again, will look into it when I have some time.

kjkooistra-youwe commented:
@lukmzig I'm unable to reproduce the error. I would surely have noticed if creating/opening a definition would have caused an error. I updated the 11.x branch to the latest version. Can you provide more details on how to reproduce the issue? Or a trace where the trim is called?

lukmzig commented:
[](url)Hi @kjkooistra-youwe, sure, seems like the exception is coming from [here](https://github.com/pimcore/pimcore/pull/14454/files#diff-bbea58ca69510845c96706b8467d7891cffb0c792866d6be1f6e258d8d9e0229R398-R399) and it is thrown on definition load by call to the `admin/class/get-select-options-usages`

Trace:
in /var/www/html/dev/pimcore/pimcore/models/DataObject/SelectOptions/Definition.php:398
0 /var/www/html/dev/pimcore/pimcore/models/DataObject/SelectOptions/Definition.php(398): trim(NULL, '\\')
1 /var/www/html/dev/pimcore/pimcore/models/DataObject/SelectOptions/Definition.php(384): Pimcore\Model\DataObject\SelectOptions\Definition->isConfiguredAsOptionsProvider(Object(Pimcore\Model\DataObject\ClassDefinition\Data\TargetGroup))
2 /var/www/html/dev/pimcore/pimcore/models/DataObject/SelectOptions/Definition.php(354): Pimcore\Model\DataObject\SelectOptions\Definition->getFieldsUsedInClass(Object(Pimcore\Model\DataObject\ClassDefinition), 'Class')

I did not do anything special except pulling your PR and adding a new definition. I also checked with my colleagues and they were able to reproduce this error as well. Let me know if this helped you or if we missed something.

kjkooistra-youwe commented:
@lukmzig Thanks. That makes sense, as those methods may return null. I'm not sure why the error didn't occur for me.

kjkooistra-youwe commented:
@lukmzig I updated the pull request, but the static analysis giving some errors now. I'm not sure where those are coming from though.

kjkooistra-youwe commented:
Fixed errors caused by changes in this commit: https://github.com/pimcore/pimcore/commit/959939066166a85668e29c23460a8e21055a2091
Fixed merge conflicts.

kjkooistra-youwe commented:
@lukmzig As this pull request keeps gathering merge conflicts, can you give any indication on when it could be merged? I'd rather not keep fixing them indefinitely.

lukmzig commented:
> @lukmzig As this pull request keeps gathering merge conflicts, can you give any indication on when it could be merged? I'd rather not keep fixing them indefinitely.

Hi @kjkooistra-youwe, thank you for reaching out. Did you by any chance already have time to have a look at the improvements which were mentioned up? It would be quite important for us that they are resolved before we could merge your improvement. 

In general, I am afraid we will not be able to merge this improvement before Pimcore 11 is released. For this reason, it is also not necessary that you resolve all conflicts right away.

kjkooistra-youwe commented:
@lukmzig Apart from the localization, the improvements were already implemented. Sorry for not indicating that more clearly.

aryaantony92 commented:
@kjkooistra-youwe Could you please resolve the conflicts? Also, the admin bundle related changes needs to be refactored into the admin-ui-bundle https://github.com/pimcore/admin-ui-classic-bundle. I can work on these, if you are busy. Thanks!

kjkooistra-youwe commented:
@aryaantony92 I'll see if I can find some time. I would appreciate it if you could implement the admin related changes.

aryaantony92 commented:
> @aryaantony92 I'll see if I can find some time. I would appreciate it if you could implement the admin related changes.

Sure, I will do that.

kjkooistra-youwe commented:
@aryaantony92 The conflicts have been resolved, although I may have to do some updates based on the automated analysis. Obviously I haven't been able to test it yet, since the admin changes still have to be applied. I also renamed the 'select-options' setting to 'select_options' to be consistent with the other settings (Kernel.php and Configuration.php).

[Original commit for reference](https://github.com/kjkooistra-youwe/pimcore/commit/86659db894f8fca0c0c31b7e9cb19d11cae3d0ba)

Conflict overview:
```
deleted by us:   bundles/AdminBundle/public/css/icons.css
deleted by us:   bundles/AdminBundle/public/js/pimcore/layout/toolbar.js
deleted by us:   bundles/AdminBundle/public/js/pimcore/object/classes/data/multiselect.js
deleted by us:   bundles/AdminBundle/public/js/pimcore/object/classes/data/select.js
deleted by us:   bundles/AdminBundle/public/js/pimcore/object/helpers/grid.js
deleted by us:   bundles/AdminBundle/public/js/pimcore/object/tags/select.js
deleted by us:   bundles/AdminBundle/src/Controller/Admin/DataObject/ClassController.php
deleted by us:   bundles/AdminBundle/src/Controller/Admin/IndexController.php
deleted by us:   bundles/AdminBundle/templates/admin/index/index.html.twig
deleted by us:   bundles/AdminBundle/translations/admin.en.yaml
deleted by us:   bundles/AdminBundle/translations/admin_ext.en.yaml
both modified:   bundles/CoreBundle/src/Command/ClassesDefinitionsBuildCommand.php
both modified:   bundles/CoreBundle/src/DependencyInjection/Configuration.php
both modified:   doc/21_Deployment/03_Configuration_Environments.md
deleted by us:   lib/Event/AdminEvents.php
added by them:   lib/Extension/Bundle/Exception/helpers/reservedWords.js
added by them:   lib/Extension/Bundle/Exception/helpers/selectField.js
added by them:   lib/Extension/Bundle/Exception/selectoptions.js
added by them:   lib/Extension/Bundle/Exception/selectoptionsitems/definition.js
both modified:   lib/Kernel.php
both modified:   models/DataObject/ClassDefinition/Data/Multiselect.php
both modified:   models/DataObject/ClassDefinition/Data/Select.php
both modified:   tests/_bootstrap.php
```

I'm not sure where why the 'lib/Extension/Bundle/Exception/*' files are in there.

aryaantony92 commented:
> @aryaantony92 The conflicts have been resolved, although I may have to do some updates based on the automated analysis. Obviously I haven't been able to test it yet, since the admin changes still have to be applied. I also renamed the 'select-options' setting to 'select_options' to be consistent with the other settings (Kernel.php and Configuration.php).
> 
> [Original commit for reference](https://github.com/kjkooistra-youwe/pimcore/commit/86659db894f8fca0c0c31b7e9cb19d11cae3d0ba)
> 
> Conflict overview:
> 
> ```
> deleted by us:   bundles/AdminBundle/public/css/icons.css
> deleted by us:   bundles/AdminBundle/public/js/pimcore/layout/toolbar.js
> deleted by us:   bundles/AdminBundle/public/js/pimcore/object/classes/data/multiselect.js
> deleted by us:   bundles/AdminBundle/public/js/pimcore/object/classes/data/select.js
> deleted by us:   bundles/AdminBundle/public/js/pimcore/object/helpers/grid.js
> deleted by us:   bundles/AdminBundle/public/js/pimcore/object/tags/select.js
> deleted by us:   bundles/AdminBundle/src/Controller/Admin/DataObject/ClassController.php
> deleted by us:   bundles/AdminBundle/src/Controller/Admin/IndexController.php
> deleted by us:   bundles/AdminBundle/templates/admin/index/index.html.twig
> deleted by us:   bundles/AdminBundle/translations/admin.en.yaml
> deleted by us:   bundles/AdminBundle/translations/admin_ext.en.yaml
> both modified:   bundles/CoreBundle/src/Command/ClassesDefinitionsBuildCommand.php
> both modified:   bundles/CoreBundle/src/DependencyInjection/Configuration.php
> both modified:   doc/21_Deployment/03_Configuration_Environments.md
> deleted by us:   lib/Event/AdminEvents.php
> added by them:   lib/Extension/Bundle/Exception/helpers/reservedWords.js
> added by them:   lib/Extension/Bundle/Exception/helpers/selectField.js
> added by them:   lib/Extension/Bundle/Exception/selectoptions.js
> added by them:   lib/Extension/Bundle/Exception/selectoptionsitems/definition.js
> both modified:   lib/Kernel.php
> both modified:   models/DataObject/ClassDefinition/Data/Multiselect.php
> both modified:   models/DataObject/ClassDefinition/Data/Select.php
> both modified:   tests/_bootstrap.php
> ```
> 
> I'm not sure where why the 'lib/Extension/Bundle/Exception/*' files are in there.

@kjkooistra-youwe Thanks for resolving the conflicts. I will take care of the admin bundle changes and will create a PR in admin-ui-bundle.

aryaantony92 commented on models/DataObject/SelectOptions/Config/Dao.php:
```diff
@@ -0,0 +1,160 @@
+        parent::configure([
+            'containerConfig' => $config['objects']['select_options']['definitions'],
+            'settingsStoreScope' => 'pimcore_select_options',
+            'storageDirectory' => $storageConfig,
```
```suggestion
            'storageConfig' => $storageConfig,
```

aryaantony92 commented on models/DataObject/SelectOptions/Config/Dao.php:
```diff
@@ -0,0 +1,160 @@
+    {
+        $config = Config::getSystemConfiguration();
+
+        $storageConfig = $config['config_location']['select_options'];
```
'select_options' can be added as a constant like https://github.com/pimcore/pimcore/blob/20cfe5c3685bf7c82af2011062a257d73c32e986/models/Asset/Image/Thumbnail/Config/Dao.php#L28, so that it will be more consistent.

kjkooistra-youwe replied:
Ah, yes. The static analysis complained about an unused constant, so I removed it. I'll put it back.

aryaantony92 commented:
@kjkooistra-youwe I have created a PR for admin-ui-bundle https://github.com/pimcore/admin-ui-classic-bundle/pull/291. Could you please sign the CLA there https://github.com/pimcore/admin-ui-classic-bundle/pull/291#issuecomment-1733585637 ? Thank you!

kjkooistra-youwe commented:
I noticed similar settings are also configured in .github/ci/files/config/config.yaml. Should it be updated as well?

aryaantony92 commented:
> I noticed similar settings are also configured in .github/ci/files/config/config.yaml. Should it be updated as well?

Yes, this config is related to the codeception tests. You can add it there too to make it consistent with others. Thanks!

aryaantony92 commented:
> > I noticed similar settings are also configured in .github/ci/files/config/config.yaml. Should it be updated as well?
> 
> Yes, this config is related to the codeception tests. You can add it there too to make it consistent with others. Thanks!

@kjkooistra-youwe Also, we can remove this file https://github.com/pimcore/pimcore/pull/14454/files#diff-f43ff01c1af018b3ee31fc731c951f862bf7024e87813487dc43090c24105b2b as we already moved it to thje admin-ui bundle

aryaantony92 commented:
@kjkooistra-youwe 

- [ ] ![image](https://github.com/pimcore/pimcore/assets/97134765/9e65eee0-5c43-423b-9810-8f7014e2ec05)
It would be nice to change this option label from 'Class / Service' to 'Options Provider'. 

- [ ] ![image](https://github.com/pimcore/pimcore/assets/97134765/72039ead-9e31-49cd-9795-3ab980869362)
Could you please avoid the prefix 'Class' here so that it would be consistent with others (Field Collections)

It's also fine to add those in a follow up PR.

kjkooistra-youwe commented:
@aryaantony92 

You can change the label in your PR https://github.com/pimcore/admin-ui-classic-bundle/pull/291/files#diff-b7044d81c8cbf9664c08384185ac516051a9235b8a234325bf22993531c1fc6b

The Class prefix was added on purpose to distinguish between field collections and classification store, for example. Otherwise it might become confusing when the same name is used.

aryaantony92 commented:
@kjkooistra-youwe Could you please readd this file https://github.com/pimcore/admin-ui-classic-bundle/blob/2970bebcc20f21ccb0a5caa2110b0f9dbfb5fa20/src/OptionsProvider/SelectOptionsOptionsProvider.php to the core bundle https://github.com/pimcore/pimcore/tree/11.x/bundles/CoreBundle/src as the tests are failing and it should actually belong to the corebundle.

dvesh3 commented on doc/21_Deployment/03_Configuration_Environments.md:
```diff
@@ -91,6 +91,11 @@ pimcore:
                 directory: '/var/www/html/var/config/object_custom_layouts'
+        select_options:
+            write_target:
+                type: 'symfony-config'
+            options:
+                directory: '/var/www/html/var/config/select_options'
```
```suggestion
                type: 'symfony-config'
                options:
                    directory: '/var/www/html/var/config/select_options'
```

dvesh3 commented on lib/Kernel.php:
```diff
@@ -28,6 +28,7 @@
 use Pimcore\Config\LocationAwareConfigRepository;
 use Pimcore\Event\SystemEvents;
 use Pimcore\HttpKernel\BundleCollection\BundleCollection;
+use Pimcore\Model\WebsiteSetting;
```
```suggestion
```

kjkooistra-youwe replied:
Oops

dvesh3 commented on models/DataObject/SelectOptions/Config.php:
```diff
@@ -0,0 +1,377 @@
+    {
+        $reservedWordsHelper = new ReservedWordsHelper();
+        if ($reservedWordsHelper->isReservedWord($id)) {
+            throw new \Exception(
```
```suggestion
            throw new \InvalidArgumentException(
```
throw specific

dvesh3 commented on models/DataObject/SelectOptions/Config.php:
```diff
@@ -0,0 +1,377 @@
+        try {
+            $selectOptions = RuntimeCache::get($cacheKey);
+            if (!$selectOptions instanceof self) {
+                throw new \Exception('Select options in registry is invalid', 1678353750987);
```
```suggestion
                throw new \RuntimeException('Select options in registry is invalid', 1678353750987);
```

dvesh3 commented on models/DataObject/SelectOptions/Config.php:
```diff
@@ -0,0 +1,377 @@
+        // Check whether ID is available
+        $id = $data[static::PROPERTY_ID] ?? null;
+        if (empty($id)) {
+            throw new \Exception('ID is mandatory for select options definition', 1676646778230);
```
```suggestion
            throw new \RuntimeException('ID is mandatory for select options definition', 1676646778230);
```

dvesh3 commented on models/DataObject/SelectOptions/Config/Dao.php:
```diff
@@ -0,0 +1,162 @@
+    {
+        $id = $this->model->getId();
+        if (empty($id)) {
+            throw new \Exception('A select options definition needs an ID to be saved!', 1676639722696);
```
```suggestion
            throw new \InvalidArgumentException('A select options definition needs an ID to be saved!', 1676639722696);
```

dvesh3 commented on models/DataObject/SelectOptions/Config/Dao.php:
```diff
@@ -0,0 +1,162 @@
+        }
+
+        if (!preg_match('/[A-Z][a-zA-Z0-9]+/', $id)) {
+            throw new \Exception('Invalid ID: Must start with capital letter, followed by alphanumeric characters', 1676639634486);
```
```suggestion
            throw new \InvalidArgumentException('Invalid ID: Must start with capital letter, followed by alphanumeric characters', 1676639634486);
```

dvesh3 commented on models/DataObject/SelectOptions/Config/Dao.php:
```diff
@@ -0,0 +1,162 @@
+            $report[] = $className . ': ' . implode(', ', $fieldNames);
+        }
+
+        throw new \Exception(
```
```suggestion
        throw new \RuntimeException(
```

dvesh3 commented on models/DataObject/SelectOptions/Config/Listing.php:
```diff
@@ -0,0 +1,89 @@
+class Listing extends Model\AbstractModel implements \IteratorAggregate
+{
+    /**
+     * @internal
```
why internal?

kjkooistra-youwe replied:
I think I copied it from the thumbnail setup.

dvesh3 replied:
I'm not sure why it is like this in Thumbnail listing. Let's keep it then for consistency.

dvesh3 commented:
@kjkooistra-youwe great work :clap: It will be a highlight of v11.1. Thank you very much for co-operation as this took a bit time to review :wink:

kjkooistra-youwe commented:
Also thanks to @aryaantony92, who migrated the interface changes to the admin-ui bundle.
